Generating type-safe constructs from Minecraft's data

foaw opened the pull request:
# Description

This pull request should close the issue about generating type-safe structures from [Minecraft's data](https://wiki.vg/Data_Generators) #11.

foaw committed:
feat(data-gen): generate `struct Block`

foaw committed:
feat(data-gen): generate `impl Default for Block`

foaw committed:
feat(data-gen): sanitise property & variant names

foaw committed:
feat(data-gen): generate `struct BlockProperty`

foaw committed:
feat(data-gen): generate `impl From<u32> for Block`

foaw committed:
refactor(data-gen): use io::Write as IoWrite

foaw committed:
refactor(data-gen): split code generation

foaw committed:
refactor(data-gen): import traits without bringing items into scope

foaw committed:
feat(data-gen): generate `impl From<u8> for BlockProperty`

foaw committed:
feat(data-gen): generate `impl From<bool> for BlockProperty`

JarvisCraft commented on crates/data-generator/Cargo.toml:
```diff
@@ -0,0 +1,22 @@
+[package]
+name = "rjacraft_data_generator"
+description = "Data generator for Rjacraft."
+categories = ["codegen"]
```
You should use an existing category: https://crates.io/categories

foaw replied:
Done; see tree at [`cd92b18`](https://github.com/rjacraft/rjacraft/blob/cd92b18f2664e0a5f18cdae4da968e41905e32bf/crates/data-generator/Cargo.toml#L4).
https://github.com/rjacraft/rjacraft/blob/cd92b18f2664e0a5f18cdae4da968e41905e32bf/crates/data-generator/Cargo.toml#L4

JarvisCraft commented on crates/data-generator/src/blocks/mod.rs:
```diff
@@ -0,0 +1,65 @@
+use self::model::{Block, BlockProperty};
+use self::output::{BlockConvert, BlockDefault, BlockStruct, PropertyConvert, PropertyStruct};
+
+pub(crate) use json::ParseError as JsonParseError;
```
Please use `self::` for less ambiguity and better automatic formatting.

golf25565 replied:
```suggestion
pub use json::ParseError as JsonParseError;
```
Unnecessary encapsulation

foaw replied:
> Unnecessary encapsulation

Done; see tree at 8fe9b7bac381e73e761d8d831a82a03d5096b4d3:
https://github.com/rjacraft/rjacraft/blob/ab3ad367651e2a16b94fc91e1502b87481701c8a/crates/data-generator/src/blocks/mod.rs#L3

JarvisCraft commented on crates/types/build.rs:
```diff
@@ -0,0 +1,23 @@
+        .open(Path::new(&out_dir).join("blocks.rs"))
+        .expect("create destination file");
+
+    let src_path = Path::new("../../mc-data/blocks.json");
```
This won't work in most cases since when published only files relative to the crate root will be included. You should move the data JSONs closer, i.e. to a directory on the crate root.

foaw replied:
Can we work around this with Cargo's [`package.include`](https://doc.rust-lang.org/cargo/reference/manifest.html#the-exclude-and-include-fields)? If not, what do you think would be the lesser evil here, including the string into the final binary at build time, or moving `blocks.json` inside of `crates/type`?

I don't like the idea of moving it anywhere, as it plainly the natural course of action to keep all resources in one place, which `contrib/` could end up becoming, as opposed to having two or three of them. Also, this is going to be especially important for users as I see it.

foaw replied:
So, I tried `package.include` with `"../../contrib/*"`, but it didn't seem to work–`cargo package --list` still only reported the `crates/types` files.

While trying to get this to work I thought of a different approach, to work in reverse: that is, symlink not `/crates/types/blocks.json` to `/contrib/blocks.json`, but the other way around.

JarvisCraft replied:
Just move `./contrib/` to `./crates/types/contrib/`

JarvisCraft commented on crates/types/Cargo.toml:
```diff
@@ -0,0 +1,14 @@
+repository.workspace = true
+
+[build-dependencies]
+rjacraft_data_generator = { path = "../data-generator" }
```
Как и прочие зависимости, объяви эту в корневом `Cargo.toml` рядом (и не забудь указать версию)

foaw replied:
Done; see [`cd92b18:/crates/types/Cargo.toml`](https://github.com/rjacraft/rjacraft/blob/cd92b18f2664e0a5f18cdae4da968e41905e32bf/crates/types/Cargo.toml) & [`cd92b18:/Cargo.toml`](https://github.com/rjacraft/rjacraft/blob/cd92b18f2664e0a5f18cdae4da968e41905e32bf/Cargo.toml).

JarvisCraft commented on crates/types/build.rs:
```diff
@@ -0,0 +1,23 @@
+extern crate rjacraft_data_generator;
```
There is no reason to use `extern crate` here.

foaw replied:
Figured it would provide more context, even though it's not mandatory it newer versions of Rust. Will remove.

foaw replied:
Done; see [`cd92b18:/crates/types/build.rs`](https://github.com/rjacraft/rjacraft/blob/cd92b18f2664e0a5f18cdae4da968e41905e32bf/crates/types/build.rs).

JarvisCraft commented on crates/types/build.rs:
```diff
@@ -0,0 +1,23 @@
+extern crate rjacraft_data_generator;
+
+use rjacraft_data_generator::*;
```
Since you already use a few items from the package, either import them individually or write the full path.

foaw replied:
Yeah, apparently I forgot to remove the usage. The crate was originally built for testing purposes, rather than to stay as a permanent solution.

foaw replied:
Done; see [`cd92b18:/crates/types/build.rs`](https://github.com/rjacraft/rjacraft/blob/cd92b18f2664e0a5f18cdae4da968e41905e32bf/crates/types/build.rs).

JarvisCraft commented on crates/types/src/lib.rs:
```diff
@@ -0,0 +1,12 @@
+}
+
+#[cfg(test)]
+mod tests {
```
Remove the default example.

foaw replied:
No longer relevant I believe.

JarvisCraft commented on crates/types/build.rs:
```diff
@@ -0,0 +1,23 @@
+    let src_path = Path::new("../../mc-data/blocks.json");
+    let json_data = std::fs::read_to_string(&src_path).expect("read input JSON");
+
+    gen_structs(json_data, &mut dest_file).expect("generate code");
```
Either the function should be renamed to `gen_block_structs` or the output file should not be named `blocks.rs`

foaw replied:
Addressed this earlier. WDYT we should name it, `gen_code` perhaps?

I'm all in favour of getting rid of `blocks.json` as hardcoding the path, but then again, we haven't even settled on the data source yet.

JarvisCraft replied:
Separate methods (for blocks, for dimmensions etc) are still useful. What I don't like is a method which works *exactly* with blocks but does not indicate it.

foaw replied:
Okay, let's rename it for now, though I expect this to change soon.

foaw replied:
Just to make sure we're on the same page. This is the method that just goes through unexported modules and generates code for their kind–be it blocks, potions, or whatever; while it's now used exclusively for blocks, it won't stay this way forever. Are you sure we want to rename it? @JarvisCraft

JarvisCraft replied:
I do think so since it explicitly reads `blcoks.json`. Once support for the other structures is addded, the signature will change and thus the name.

JarvisCraft commented on crates/data-generator/src/blocks/json.rs:
```diff
@@ -0,0 +1,132 @@
+    Ok(result)
+}
+
+fn consume_until_colon<'a>(input: &'a str) -> &'a str {
```
There is no need for explicit lifetimes here.

foaw replied:
Done; see [`cd92b18`](https://github.com/rjacraft/rjacraft/blob/cd92b18f2664e0a5f18cdae4da968e41905e32bf/crates/data-generator/src/blocks/json.rs#L130).
https://github.com/rjacraft/rjacraft/blob/cd92b18f2664e0a5f18cdae4da968e41905e32bf/crates/data-generator/src/blocks/json.rs#L130

JarvisCraft commented on crates/data-generator/src/blocks/json.rs:
```diff
@@ -0,0 +1,132 @@
+use super::model::{Block, BlockProperty, BlockState, BlockStateProperty, Id};
+
+#[derive(Debug, Deserialize)]
+struct SerdeId(u32);
```
This is quite a non-descriptive name. Either just use `Id` if it is generally applicable here, or add a more domain-oriented prefix.

Also, is there any need for this newtype at all?

foaw replied:
Not really I guess, not any more. There was once a single model, which was split up into two later on. Now I don't think there is. Speaking of practical application–have a look at `impl ToTokens for models::Id` in `output.rs`, this alone should already give you chills.

JarvisCraft commented on crates/data-generator/src/blocks/mod.rs:
```diff
@@ -0,0 +1,65 @@
+}
+
+fn gen_block_code(block: &Block, stream: &mut TokenStream) {
+    let block_struct: BlockStruct = BlockStruct::from(block);
```
```suggestion
    let block_struct = BlockStruct::from(block);
```

foaw replied:
I don't think this is still relevant as of now. Correct me if I'm wrong.
https://github.com/rjacraft/rjacraft/blob/feature/data-generator/crates/data-generator/src/blocks/mod.rs#L19-L29

JarvisCraft commented on crates/data-generator/src/blocks/model.rs:
```diff
@@ -0,0 +1,30 @@
+
+#[derive(Debug)]
+pub struct Block {
+    pub name: String, // PascalCase, dirty
```
Place the comments on top of the items

foaw replied:
I'd rather remove them as they do more harm than good for what they are.

foaw replied:
Outdated for this struct as of cd92b18f2664e0a5f18cdae4da968e41905e32bf. This is still a thing in other places though, like here:

https://github.com/rjacraft/rjacraft/blob/cd92b18f2664e0a5f18cdae4da968e41905e32bf/crates/data-generator/src/blocks/output.rs#L188-L191

Are you finding this to be an issue just for the public items, or generally for all comments that stand to the right-hand side of declarations? If so, I'd like to hear another alternative, as the docs are clumsy, and I don't even consider this as documentation of any sort–I only left them there for clarity's sake.

JarvisCraft commented on crates/data-generator/src/blocks/model.rs:
```diff
@@ -0,0 +1,30 @@
+pub struct BlockProperty {
+    pub block_name: String,    // PascalCase, dirty
+    pub prop_name: String,     // snake_case, dirty
+    pub variants: Vec<String>, // PascalCase, dirty
```
Just don't use table alignment in source code

foaw replied:
FYI, this is how Rustfmt handles such comments. Although if have them to the right side, then better aligned than not.

JarvisCraft commented on crates/data-generator/src/blocks/json.rs:
```diff
@@ -0,0 +1,132 @@
+    Json(#[from] serde_json::Error),
+}
+
+pub(super) fn parse_block_registry(json_data: String) -> Result<Vec<Block>, ParseError> {
```
I would prefer us to use `impl Read` here and in all callers.

`serde_json` supports it natively: https://docs.rs/serde_json/latest/serde_json/fn.from_reader.html

foaw replied:
Done; see [`d8fa342`](https://github.com/rjacraft/rjacraft/blob/d8fa342630650f9e78acc4219e82bd18e281f618/crates/data-generator/src/blocks/json.rs#L34).
https://github.com/rjacraft/rjacraft/blob/d8fa342630650f9e78acc4219e82bd18e281f618/crates/data-generator/src/blocks/json.rs#L34

JarvisCraft commented on crates/data-generator/src/blocks/json.rs:
```diff
@@ -0,0 +1,132 @@
+    ) -> BlockProperty {
+        let prop_variants = prop_variants
+            .iter()
+            .map(|prop_variant| transform_to_pascal_case(prop_variant))
```
```suggestion
            .map(transform_to_pascal_case)
```

The same in all similar places

foaw replied:
Should be done; look in worktree at [`d8fa342`](https://github.com/rjacraft/rjacraft/tree/feature/data-generator). Quickly skimming through the code base, I didn't find anything.

JarvisCraft commented on crates/data-generator/src/blocks/json.rs:
```diff
@@ -0,0 +1,132 @@
+    }
+
+    fn block_state_property(prop_name_sc: String, prop_variant: &String) -> BlockStateProperty {
+        let prop_variant_pc = transform_to_pascal_case(&prop_variant);
```
IMO, the code would be more readable without this extra local variables.

foaw replied:
Obsolete as of d8fa342630650f9e78acc4219e82bd18e281f618.

JarvisCraft commented on crates/data-generator/src/blocks/mod.rs:
```diff
@@ -0,0 +1,65 @@
+
+        let syn_tree = parse2(stream).expect("parse TokenStream into syn::File");
+        let pretty_output = prettyplease::unparse(&syn_tree);
+        writeln!(sink, "{}", pretty_output).expect("write to destination file");
```
Since write may fail, you should return a error instead of panicking.

foaw replied:
Done; see in tree at [`d8fa342`](https://github.com/rjacraft/rjacraft/blob/d8fa342630650f9e78acc4219e82bd18e281f618/crates/data-generator/src/blocks/mod.rs#L28).

JarvisCraft commented on crates/data-generator/src/lib.rs:
```diff
@@ -0,0 +1,17 @@
+use std::io::Write;
+
+use thiserror::Error;
```
It's better to use full path in dervive, since this collides with the name of an std trait.

foaw replied:
Done; see in tree at [`91c5aa3`](https://github.com/rjacraft/rjacraft/blob/91c5aa38e29a0dee9b8036c73595e2e9d889eafd/crates/data-generator/src/lib.rs#L6).

foaw committed:
feat(data-gen): use `Block` inside of block modules

foaw committed:
feat(data-gen): derive `Default` for `Block`

foaw committed:
feat(data-gen): generate `impl Default for Property`

foaw committed:
feat(data-gen): generate `impl Not for BoolProperty`

foaw committed:
refactor(data-gen): generate into modules

foaw committed:
refactor(data-gen): to_ident() extension method

JarvisCraft commented on crates/data-generator/src/lib.rs:
```diff
@@ -0,0 +1,17 @@
+pub enum GenerateError {
+    #[error("I/O error: {0}")]
+    Io(#[from] std::io::Error),
+    #[error("error generating blocks: {0}")]
```
```suggestion
    #[error("an I/O operation failed")]
    Io(#[from] std::io::Error),
    #[error("failed to generate blocks")]
```

foaw replied:
See https://github.com/rjacraft/rjacraft/pull/14#discussion_r1322288625.

foaw commented on crates/data-generator/src/blocks/json.rs:
```diff
@@ -0,0 +1,194 @@
+                let var_name = transform_to_pascal_case(&prop_variant);
+                let var_name = defuse_variant_name(var_name);
+
+                match (all_bools, all_nums) {
+                    (true, false) => {
+                        let value = prop_variant.parse().expect(&prop_variant);
+                        BlockPropertyVariant::Bool(value, var_name)
+                    }
+                    (false, true) => {
+                        let ord = prop_variant.parse().unwrap();
+                        BlockPropertyVariant::Numeric(ord, var_name)
+                    }
+                    (_, _) => BlockPropertyVariant::Regular(var_name),
+                }
```
Replace with a usual if-else-if ladder, or instead of `bool`s store a `fn(var_name, prop_variant) -> BlockPropertyVariant`, and then call it from here regardless of which variant it generates.

foaw commented on crates/data-generator/src/blocks/output.rs:
```diff
@@ -0,0 +1,628 @@
+        }
+    }
+
+    impl ToTokens for Id {
+        fn to_tokens(&self, tokens: &mut TokenStream) {
+            tokens.append(Literal::u32_unsuffixed(self.0))
+        }
+    }
```
This looks out of place here. Moving the implementation elsewhere won't help a whole lot; doing `Literal::u32_unsuffixed` at every call site isn't any better either.

golf25565 replied:
`u32` implements `ToTokens` too so maybe use that

foaw replied:
Unlike us, they make suffixed literals, that is, with `u32` suffix at the end.

golf25565 replied:
And so what? That *is* our protocol ID type, no?

JarvisCraft commented on crates/data-generator/Cargo.toml:
```diff
@@ -0,0 +1,22 @@
+thiserror.workspace = true
+serde.workspace = true
+serde_json.workspace = true
+syn = { workspace = true, features = ["parsing"] }
```
This already is a default feature, and you use more of these

foaw replied:
Could you elaborate on this? In the parent [`Cargo.toml`](https://github.com/rjacraft/rjacraft/blob/91c5aa38e29a0dee9b8036c73595e2e9d889eafd/Cargo.toml#L31), `syn` has default features disabled. Am I missing something?

JarvisCraft commented on crates/data-generator/src/blocks/json.rs:
```diff
@@ -0,0 +1,194 @@
+
+#[derive(Debug, Error)]
+pub enum ParseError {
+    #[error("malformed JSON data: {0}")]
```
```suggestion
    #[error("malformed JSON data")]
```

foaw replied:
Done; see in tree at [`91c5aa3`](https://github.com/rjacraft/rjacraft/blob/91c5aa38e29a0dee9b8036c73595e2e9d889eafd/crates/data-generator/src/blocks/json.rs#L28-L32).

JarvisCraft commented on crates/data-generator/src/blocks/json.rs:
```diff
@@ -0,0 +1,194 @@
+            let def_block_state = block_states
+                .iter()
+                .find(|(_, v)| v.default)
+                .map(|(_, v)| v.clone())
```
```suggestion
            let def_block_state = block_states
                .values()
                .find(|v| v.default)
                .map(|v| v.clone())
```

JarvisCraft commented on crates/data-generator/src/blocks/json.rs:
```diff
@@ -0,0 +1,194 @@
+                .iter()
+                .find(|(_, v)| v.default)
+                .map(|(_, v)| v.clone())
+                .expect("default state exists");
```
IMO, such errors should be reported via `Result`

foaw replied:
Obsolete as of 91c5aa38e29a0dee9b8036c73595e2e9d889eafd.

JarvisCraft commented on crates/data-generator/src/blocks/json.rs:
```diff
@@ -0,0 +1,194 @@
+    &input[input.find(':').map_or(0, |pos| pos + 1)..]
+}
+
+fn transform_to_pascal_case(input: impl AsRef<str>) -> String {
```
`transform_` is a really weird prefix when we have just `to_`

foaw replied:
Obsolete as of 91c5aa38e29a0dee9b8036c73595e2e9d889eafd. See `name.rs`:
https://github.com/rjacraft/rjacraft/blob/91c5aa38e29a0dee9b8036c73595e2e9d889eafd/crates/data-generator/src/name.rs#L19-L27

foaw commented on crates/data-generator/src/blocks/output.rs:
```diff
@@ -0,0 +1,628 @@
+        }
+    }
+
+    pub struct BlockStructField {
+        pub prop_name: Ident,      // honey_level
+        pub prop_enum_name: Ident, // HoneyLevel
+    }
```
These comments all over the place only made it harder to comprehend. One of the solutions I see is to have names of distinct things as `struct`s that encapsulate the handling details as an implementation detail.

For example, here we could get by with a single `PropertyName`, which we'd use like this:
```rs
impl ToTokens for BlockStructField {
    fn to_tokens(&self, tokens: &mut TokenStream) {
        // This naming is lame but we can figure something out later.
        let field_name = self.prop_name.field_name();
        let enum_name = self.prop_name.enum_name();

        // "pub open: Open,"
        tokens.extend(quote!(pub #field_name: #enum_name,))
    }
}
```

foaw replied:
On top of that, this particular struct is probably not worth being extracted from `BlockStruct`.

The guideline for me was that functionality taking more than 5-10 lines should be extracted into a separate struct, which (1) proved too short to be extracted, and (2) no longer the case after all the changes to the original code of `BlockStructField`.

golf25565 replied:
`crate::name::Name`

JarvisCraft commented on crates/data-generator/src/blocks/json.rs:
```diff
@@ -0,0 +1,194 @@
+
+fn all_bools(variants: impl AsRef<[String]>) -> bool {
+    const TF: [&str; 2] = ["true", "false"];
+    const FT: [&str; 2] = ["false", "true"];
```
There is no need for these constants

foaw replied:
Obsolete as of 91c5aa38e29a0dee9b8036c73595e2e9d889eafd. See `json.rs`:
https://github.com/rjacraft/rjacraft/blob/942a85d3957b177e44eb0d87f6718f0b0e0e1666/crates/data-generator/src/blocks/json.rs#L140-L153

JarvisCraft commented on crates/data-generator/src/blocks/json.rs:
```diff
@@ -0,0 +1,194 @@
+
+fn all_nums(variants: impl IntoIterator<Item = impl AsRef<str>>) -> bool {
+    variants
+        .into_iter()
+        .all(|s| s.as_ref().chars().all(|c| c.is_numeric()))
```
```suggestion
        .iter()
        .all(|s| s.chars().all(|c| c.is_numeric()))
```

JarvisCraft commented on crates/data-generator/src/blocks/json.rs:
```diff
@@ -0,0 +1,194 @@
+
+fn defuse_property_name(property_name: impl AsRef<str>) -> String {
+    match property_name.as_ref() {
+        "type" => "kind".to_string(),
```
We've decided to use `r#`-identifiers for this purpose.

JarvisCraft commented on crates/data-generator/src/blocks/model.rs:
```diff
@@ -0,0 +1,59 @@
+        self.variants
+            .iter()
+            .next()
+            .map(|x| matches!(x, BlockPropertyVariant::Numeric(_, _)))
```
Should not we check that *all* variants are numeric?

foaw replied:
I can't remember what that part of the code was for. Could yo see if this is still relevant?

foaw commented on crates/data-generator/src/blocks/output.rs:
```diff
@@ -0,0 +1,628 @@
+        fn to_tokens(&self, tokens: &mut TokenStream) {
+            use super::{block_exports::BlockExports, block_mod::BlockMod};
+
+            let mut block_mods = TokenStream::new();
+            self.blocks
+                .iter()
+                .map(BlockMod)
+                .for_each(|block| block.to_tokens(&mut block_mods));
```
```rs
let block_mods = self.blocks.iter().map(BlockMod);
// ...
query!( #(#block_mods)* )
```

JarvisCraft commented on crates/data-generator/src/blocks/model.rs:
```diff
@@ -0,0 +1,59 @@
+#[derive(Debug, Clone)]
+pub struct Id(pub u32);
```
What is the purpose of this newtype?

JarvisCraft commented on crates/data-generator/src/blocks/model.rs:
```diff
@@ -0,0 +1,59 @@
+        match self {
+            &BlockPropertyVariant::Regular(ref s) => s,
+            &BlockPropertyVariant::Bool(_, ref s) => s,
+            &BlockPropertyVariant::Numeric(_, ref s) => s,
```
```suggestion
            BlockPropertyVariant::Regular(s) => s,
            BlockPropertyVariant::Bool(_, s) => s,
            BlockPropertyVariant::Numeric(_, s) => s,
```

foaw replied:
Obsolete as of 91c5aa38e29a0dee9b8036c73595e2e9d889eafd.

JarvisCraft commented on crates/data-generator/src/blocks/output.rs:
```diff
@@ -0,0 +1,628 @@
+            for (id, state_inst) in &self.0.id_states {
+                let state_inst = state_inst;
+
+                // "4287 => Farmland,"
```
Nope

```suggestion
```

foaw replied:
What's wrong with this?
https://github.com/rjacraft/rjacraft/blob/851b6180724238e291e0c042b2300f1200b45169/crates/data-generator/src/blocks/output.rs#L267-L272

JarvisCraft commented on crates/data-generator/src/blocks/output.rs:
```diff
@@ -0,0 +1,628 @@
+            let prop_name_pc = block_prop.name_pc.clone();
+            let ordered = block_prop.is_num();
+
+            let prop_variants: Vec<PropertyEnumVariant> = block_prop
```
```suggestion
            let prop_variants: Vec<_> = block_prop
```

even better:

```suggestion
            let prop_variants = block_prop
```

since rustc knows the type of `prop_variants` based on its following usage.

foaw replied:
Obsolete as of cd92b18f2664e0a5f18cdae4da968e41905e32bf.

JarvisCraft commented on crates/data-generator/src/blocks/output.rs:
```diff
@@ -0,0 +1,628 @@
+                    var_name: var_name.defused_name().to_string(),
+                    var_ordinal: Some(block_prop)
+                        .filter(|&prop| prop.is_num())
+                        .map(|_| var_ord),
```
This is weird way to introduce an `Option`.

Why not just:

```suggestion
                    var_ordinal: if block_prop.is_num() {
                        Some(var_ord)
                    } else {
                        None
                    },
```

Why is much easier to understand.

Even more, this is such a common use case that there is [`bool::then_some`](https://doc.rust-lang.org/std/primitive.bool.html#method.then_some):

```suggestion
                    var_ordinal: prop.is_num().then_some(var_ord),
```

foaw replied:
Obsolete for this change as of cd92b18f2664e0a5f18cdae4da968e41905e32bf; still, updated the rest of the code base to use `Option::then_some`.

JarvisCraft commented on crates/data-generator/src/blocks/output.rs:
```diff
@@ -0,0 +1,628 @@
+    }
+
+    impl ToTokens for PropertyEnumVariant {
+        fn to_tokens(&self, tokens: &mut TokenStream) {
```
Can't you make this (and the similar) methods return their own `TokenStream` which would be appended to the original stream?

foaw replied:
Could you perhaps provide an example of what this should look like?

golf25565 committed:
refactor(data-gen): revise the mid-level representation

Co-authored-by: Artem Popov <git@tema.sh>

foaw commented on crates/data-generator/src/blocks/output.rs:
```diff
@@ -0,0 +1,628 @@
+    impl ToTokens for BlockIntoU32<'_> {
+        fn to_tokens(&self, tokens: &mut TokenStream) {
+            let mut match_arms = TokenStream::new();
+            for (id, state) in &self.0.id_states {
+                // "LargeAmethystBud { Facing: North, Waterlogged: True, }"
+                let state_inst = state.to_token_stream();
+
+                // "LavaCauldron => 7402,"
+                let stream = quote! { #state_inst => #id, };
+                stream.to_tokens(&mut match_arms);
+            }
```
Must've lost the change while rebasing. This should be:
```rs
for (id, state_inst) in &self.0.id_states {
    // "RedstoneLamp { Lit: Lit::True, } => 7417,"
    let match_arm = quote! { #state_inst => #id, };
    match_arm.to_tokens(&mut match_arms);
}
```
Or more generally, though I'm not in favour of this option:
```rs
let match_arms = self.0.id_states.iter().map(|(id, state)| {
    // "RedstoneLamp { Lit: Lit::True, } => 7417,"
    quote! { #state_inst => #id, }
});
```
Also, back to `BlockStructInst`, this is roughly what it should look like after inlining `impl ToTokens` for it.

foaw commented on crates/data-generator/src/blocks/output.rs:
```diff
@@ -0,0 +1,628 @@
+        fn to_tokens(&self, tokens: &mut TokenStream) {
+            let mut match_arms = TokenStream::new();
+            for (id, state_inst) in &self.0.id_states {
+                let state_inst = state_inst;
```
This is, well, bizarre..? Same as above, anyway.

foaw commented on crates/data-generator/src/blocks/output.rs:
```diff
@@ -0,0 +1,628 @@
+    }
+}
+
+trait StringExt {
+    fn to_ident(&self) -> proc_macro2::Ident;
+}
+
+impl<T: AsRef<str>> StringExt for T {
+    fn to_ident(&self) -> proc_macro2::Ident {
+        proc_macro2::Ident::new(self.as_ref(), proc_macro2::Span::call_site())
+    }
+}
```
Ideally I'd prefer to have it in a separate module, but as it stands it's not worth another file.

foaw commented on crates/data-generator/src/lib.rs:
```diff
@@ -0,0 +1,17 @@
+
+mod blocks;
+
+#[derive(Debug, Error)]
+pub enum GenerateError {
+    #[error("I/O error: {0}")]
+    Io(#[from] std::io::Error),
+    #[error("error generating blocks: {0}")]
+    Blocks(#[from] blocks::GenerateError),
+}
```
As part of the code style unification, don't use `{0}` anywhere across the code base.

foaw commented on crates/types/build.rs:
```diff
@@ -0,0 +1,23 @@
+        .open(Path::new(&out_dir).join("blocks.rs"))
+        .expect("create destination file");
+
+    let src_path = Path::new("../../mc-data/blocks.json");
```
Deal with the source of data–should we download it ourselves as part of code generation, or have users bring a `blocks.json` of their own?

foaw replied:
Let's shed some light on this. What have we settled on? @rjacraft/developer

golf25565 replied:
Don't download anything and just make the path configurable

foaw commented on crates/data-generator/src/blocks/model.rs:
This whole thing is utterly abysmal and should be **demolished**. 💨

foaw commented on crates/data-generator/src/blocks/mod.rs:
```diff
@@ -0,0 +1,29 @@
+mod model;
+mod output;
+
+#[derive(Debug, Error)]
+pub enum GenerateError {
+    #[error("parsing error: {0}")]
+    Parse(#[from] JsonParseError),
+    #[error("I/O error writing: {0}")]
+    Write(#[from] IoError),
+}
```
Same as below; in short, don't use `{0}`.

JarvisCraft commented on crates/data-generator/src/blocks/output.rs:
```diff
@@ -0,0 +1,628 @@
+    use proc_macro2::{Ident, Literal, TokenStream};
+    use quote::{quote, ToTokens};
+
+    use super::StringExt as _;
```
Do we actually need to hide the ext-trait?

foaw replied:
Obsolete as of cd92b18f2664e0a5f18cdae4da968e41905e32bf.

JarvisCraft commented on crates/data-generator/src/blocks/output.rs:
```diff
@@ -0,0 +1,628 @@
+
+    pub struct PropertyFromU8<'a> {
+        pub enum_name: &'a Ident,
+        pub vars: &'a Vec<BlockPropertyVariant>,
```
```suggestion
        pub vars: &'a [BlockPropertyVariant],
```

I don't think that you gain anything helpful from non-`mut` ref to `Vec` instead of slice.

foaw committed:
refactor(data-gen): don't clone `Name`'s `Ident`s where possible

foaw committed:
refactor(data-gen): address the remarks

foaw commented on crates/types/Cargo.toml:
```diff
@@ -0,0 +1,14 @@
+[package]
+name = "rjacraft_types"
+description = "Types for rjacraft"
+version.workspace = true
+edition.workspace = true
+rust-version.workspace = true
+license.workspace = true
+keywords.workspace = true
+categories.workspace = true
+homepage.workspace = true
+repository.workspace = true
```
Specify the categories particular to this crate. Which ones should these be, however?

golf25565 replied:
I think `api-bindings` will be the most appropriate. The generator package can take `development-tools::build-utils`.

golf25565 replied:
As for the tags, I think that `minecraft` will suffice.

golf25565 commented on crates/data-generator/Cargo.toml:
```diff
@@ -0,0 +1,23 @@
+[package]
+name = "rjacraft_data_generator"
+description = "Data generator for Rjacraft."
+categories = ["development-tools::build-utils", "encoding"]
```
This crate doesn't need metadata

foaw replied:
Not against removing it, but why? @JarvisCraft

golf25565 replied:
Because we don't need anyone finding this crate on crates.io. It's practically junk for other developers.

JarvisCraft replied:
The crate still may be usable by other developers (although the use cases are more exotic) so I am **for** keeping metadata.

golf25565 commented on crates/data-generator/src/blocks/json.rs:
```diff
@@ -0,0 +1,153 @@
+        .all(|s| s.as_ref().chars().all(|c| c.is_numeric()))
+}
+
+fn all_bools<'a, I, II>(variants: II) -> bool
+where
+    I: AsRef<str> + ?Sized + 'a,
+    II: IntoIterator<Item = &'a I>,
+{
```
This is just hard to read, besides we're not in library code
```suggestion
fn all_bools<'a>(variants: impl IntoIterator<Item = &'a str>) -> bool {
```

golf25565 commented on crates/data-generator/src/blocks/model.rs:
```diff
@@ -0,0 +1,105 @@
+}
+
+#[derive(Debug, Clone)]
+pub enum PropertyVariant {
```
Pretty confusing name. I'd go for something like `StatePropertyValue` to make it clear that this is not what comes from the top part of the JSON object

foaw replied:
Naming is always tough. I'm sickened by the very idea of every time having to specify the name of the struct in which this struct is used as part of its name, which is arguably what `StateProperty` in `StatePropertyValue` is. I originally named it this too, but then verbosity is no longer helpful when everything is verbose.

golf25565 replied:
Then at least name it `PropertyValue`

golf25565 commented on crates/data-generator/src/blocks/model.rs:
```diff
@@ -0,0 +1,105 @@
+    Enum(Name),
+}
+
+impl PropertyVariant {
+    pub fn from(variant_raw: impl Into<String>, variants: &PropertyVariants) -> PropertyVariant {
+        let variant_raw = variant_raw.into();
+        match variants {
+            PropertyVariants::Bool => {
+                let b = variant_raw.parse().unwrap();
+                PropertyVariant::Bool(b)
+            }
+            PropertyVariants::Numeric(_) => {
+                let n = variant_raw.parse().unwrap();
+                PropertyVariant::Numeric(n)
+            }
+            PropertyVariants::Enum(_) => {
+                let var = Name::from_snake_case(variant_raw);
+                PropertyVariant::Enum(var)
+            }
+        }
+    }
+
+    pub fn variant(&self) -> Name {
+        match self {
+            Self::Enum(name) => name.clone(),
+            Self::Bool(false) => Name::new("false", "False"),
+            Self::Bool(true) => Name::new("true", "True"),
+            Self::Numeric(n) => {
+                let roman = convert_into_roman(*n);
+                Name::new(roman.to_lowercase(), roman.as_str())
+            }
+        }
+    }
+}
```
Not really model code, may be worth moving elsewhere

golf25565 commented on crates/data-generator/src/blocks/model.rs:
```diff
@@ -0,0 +1,105 @@
+    Enum(Vec<Name>),
+}
+
+impl PropertyVariants {
+    pub fn as_enum(&self) -> Vec<(Name, Option<u8>)> {
+        type One = PropertyVariant;
+        type Many = PropertyVariants;
+
+        macro_rules! one_variant {
+            ($src:expr, $one:expr, $option:expr) => {
+                $src.iter()
+                    .cloned()
+                    .map(|n| ($one(n.clone()).variant(), $option(n)))
+                    .collect::<Vec<(Name, Option<u8>)>>()
+            };
+        }
+
+        match self {
+            Many::Bool => one_variant!(vec![true, false], One::Bool, |_| None),
+            Many::Numeric(vars) => one_variant!(vars, One::Numeric, Some),
+            Many::Enum(vars) => one_variant!(vars, One::Enum, |_| None),
+        }
+    }
+}
```
Not really model code, may be worth moving elsewhere

golf25565 commented on crates/data-generator/src/blocks/output.rs:
```diff
@@ -0,0 +1,586 @@
+            });
+        }
+    }
+
+    pub struct PropertyIntoBool<'a>(&'a Ident);
+
+    impl ToTokens for PropertyIntoBool<'_> {
+        fn to_tokens(&self, tokens: &mut TokenStream) {
+            let prop_enum_name = self.0;
+            tokens.extend(quote! {
+                impl From<#prop_enum_name> for bool {
+                    fn from(value: #prop_enum_name) -> Self {
+                        value == #prop_enum_name::True
+                    }
+                }
+            });
+        }
+    }
+
+    pub struct PropertyNotBool<'a>(&'a Ident);
+
+    impl ToTokens for PropertyNotBool<'_> {
+        fn to_tokens(&self, tokens: &mut TokenStream) {
+            let prop_enum_name = self.0;
+            tokens.extend(quote! {
+                impl std::ops::Not for #prop_enum_name {
+                    type Output = Self;
+
+                    fn not(self) -> Self::Output {
+                        if self == Self::False {
+                            Self::True
+                        } else {
+                            Self::False
+                        }
+                    }
+                }
+            });
+        }
+    }
+}
```
The inline modules should be their own files. All of them.

golf25565 commented on crates/data-generator/src/blocks/output.rs:
```diff
@@ -0,0 +1,586 @@
+use super::model::Block;
+use crate::name::Name;
+
+pub(super) fn gen_blocks_mod(blocks: IndexMap<Name, Block>) -> TokenStream {
```
```suggestion
pub fn gen_blocks_mod(blocks: IndexMap<Name, Block>) -> TokenStream {
```
Unnecessary encapsulation

golf25565 commented on crates/data-generator/src/blocks/json.rs:
```diff
@@ -0,0 +1,153 @@
+    Json(#[from] serde_json::Error),
+}
+
+pub(super) fn parse_block_registry(json_data: String) -> Result<IndexMap<Name, Block>, ParseError> {
```
```suggestion
pub fn parse_block_registry(json_data: String) -> Result<IndexMap<Name, Block>, ParseError> {
```
Unnecessary encapsulation

golf25565 commented on crates/data-generator/src/blocks/mod.rs:
```diff
@@ -0,0 +1,28 @@
+    Parse(#[from] JsonParseError),
+}
+
+pub(crate) fn gen_blocks_module(
```
```suggestion
pub fn gen_blocks_module(
```
Unnecessary encapsulation

golf25565 commented on crates/data-generator/src/blocks/output.rs:
```diff
@@ -0,0 +1,586 @@
+}
+
+mod block_struct {
+    use proc_macro2::{Delimiter, Group, Ident, Punct, Spacing, TokenStream};
```
```suggestion
    use proc_macro2::*;
```
Waste of space

foaw replied:
Do we really want to introduce asterisk wildcards in Rjacraft? I don't mind wasting extra 200 bytes of disk space. @JarvisCraft

golf25565 commented on crates/data-generator/src/blocks/output.rs:
```diff
@@ -0,0 +1,586 @@
+                // "pub struct LavaCauldron;"
+                tokens.append(Punct::new(';', Spacing::Alone));
+            } else {
+                let mut stream = TokenStream::new();
+                self.properties
+                    .iter()
+                    .for_each(|prop| prop.to_tokens(&mut stream));
+
+                // "pub struct Lectern { pub has_book: HasBook, }"
+                tokens.append(Group::new(Delimiter::Brace, stream));
```
Please replace this with `quote!()` repetition

golf25565 commented on crates/data-generator/src/blocks/model.rs:
```diff
@@ -0,0 +1,105 @@
+
+#[derive(Debug, Clone)]
+pub struct States {
+    pub states: IndexMap<Id, State>,
```
Confusing name

golf25565 commented on crates/data-generator/src/blocks/output.rs:
```diff
@@ -0,0 +1,586 @@
+}
+
+mod prop_enum {
+    use proc_macro2::{Delimiter, Group, Ident, Literal, TokenStream};
```
```suggestion
    use proc_macro2::*;
```

foaw replied:
See https://github.com/rjacraft/rjacraft/pull/14#discussion_r1322305592.

golf25565 commented on crates/data-generator/src/blocks/output.rs:
```diff
@@ -0,0 +1,586 @@
+
+            let mut stream = TokenStream::new();
+            self.prop_vars.iter().for_each(|(ident, ord)| {
+                match ord {
+                    Some(ord) => quote!(#ident = #ord,),
+                    None => quote!(#ident,),
+                }
+                .to_tokens(&mut stream)
```
`quote!()` repetition can help you out here

foaw replied:
Done; see [`d51bd73:/crates/data-generator/src/blocks/output.rs`](https://github.com/rjacraft/rjacraft/blob/d51bd73625394553e9b6bc0b615eb56ce26835b7/crates/data-generator/src/blocks/output.rs#L383-L406):
https://github.com/rjacraft/rjacraft/blob/d51bd73625394553e9b6bc0b615eb56ce26835b7/crates/data-generator/src/blocks/output.rs#L383-L406

golf25565 commented on crates/data-generator/src/blocks/output.rs:
```diff
@@ -0,0 +1,586 @@
+            for &num in self.numbers.iter() {
+                let u8_num = Literal::u8_unsuffixed(num); // "4"
+                let roman_num = PropertyVariant::Numeric(num).variant();
+                let roman_num = roman_num.into_pascal_case(); // "IV"
+
+                // "4 => Self::IV,"
```
```suggestion
                let roman_num = roman_num.into_pascal_case(); // "Iv"

                // "4 => Self::Iv,"
```
Because this is pascal case

foaw replied:
For me, this is an unequivocal "no". I see where you're coming from, and I agree that we should generally adhere to Rust's standards, but we're talking Roman numerals here, not an abbreviation or whatnot. This is uncultured, indecent, and generally a slap in the face to the Romans. @JarvisCraft

golf25565 replied:
These are just names that we need to substitute numbers; who cares how they're formatted?

golf25565 commented on crates/types/Cargo.toml:
```diff
@@ -0,0 +1,14 @@
+[package]
+name = "rjacraft_types"
```
Very ambiguous naming. Could we change the name to something along the lines of `rjacraft_mc_data`? Just `rjacraft_data`? @JarvisCraft

golf25565 commented on crates/types/Cargo.toml:
```diff
@@ -0,0 +1,14 @@
+[package]
+name = "rjacraft_types"
+description = "Types for rjacraft"
```
```suggestion
description = "Rjacraft's wrapper for working with various generated protocol ID constants from Minecraft"
```

foaw replied:
Done; see [d8fa342:/crates/types/Cargo.toml](https://github.com/rjacraft/rjacraft/blob/d8fa342630650f9e78acc4219e82bd18e281f618/crates/types/Cargo.toml#L3).
https://github.com/rjacraft/rjacraft/blob/d8fa342630650f9e78acc4219e82bd18e281f618/crates/types/Cargo.toml#L3

golf25565 replied:
Still doesn't really capture the purpose of the crate. Protocol constructs go in the `protocol` crate, not in here. This is about protocol IDs and identifiers.

golf25565 commented on crates/types/build.rs:
```diff
@@ -0,0 +1,19 @@
+fn main() {
+    use std::{fs::OpenOptions, path::Path};
+
```
```suggestion
use std::{fs::OpenOptions, path::Path};

fn main() {
```

foaw replied:
Done; see [d8fa342:/crates/types/build.rs](https://github.com/rjacraft/rjacraft/blob/d8fa342630650f9e78acc4219e82bd18e281f618/crates/types/build.rs#L1).

golf25565 commented on crates/types/build.rs:
```diff
@@ -0,0 +1,19 @@
+    //println!("cargo:rerun-if-changed=../../mc-data/blocks.json");
+
+    let out_dir = std::env::var_os("OUT_DIR").expect("environment variable OUT_DIR set");
+    let mut dest_file = OpenOptions::new()
+        .write(true)
+        .create(true)
+        .truncate(true)
+        .open(Path::new(&out_dir).join("blocks.rs"))
```
Also known as `fs::File::create()`

foaw replied:
Didn't notice as it grew to basically `File::create` with the addition of `truncate(true)`. Well spotted!

foaw replied:
Done; see [d8fa342:/crates/types/build.rs](https://github.com/rjacraft/rjacraft/blob/d8fa342630650f9e78acc4219e82bd18e281f618/crates/types/build.rs#L8).

golf25565 commented on crates/types/build.rs:
```diff
@@ -0,0 +1,19 @@
+        .open(Path::new(&out_dir).join("blocks.rs"))
+        .expect("create destination file");
+
+    let src_path = Path::new("../../contrib/blocks.json");
```
Make this configurable through an env variable

foaw replied:
Done; see [d8fa342:/crates/types/build.rs](https://github.com/rjacraft/rjacraft/blob/d8fa342630650f9e78acc4219e82bd18e281f618/crates/types/build.rs#L12).

golf25565 commented on crates/types/src/lib.rs:
```diff
@@ -0,0 +1 @@
+include!(concat!(env!("OUT_DIR"), "/blocks.rs"));
```
```suggestion
pub mod blocks {
    include!(concat!(env!("OUT_DIR"), "/blocks.rs"));
}
```
The generated file shouldn't be the one to decide what module it is in.

foaw replied:
The reasoning is sound, but something about this is bugging me, as either it's unhygienic or there is simply no point in leaving the name choice to the downstream code. @JarvisCraft, WDYT?

JarvisCraft replied:
I am in favor of `blocks` module. It should not affect hygiene if implemented correctly (i.e., the generator should produce the valid results no matter where the include is located), eg., by providing extra information to it.

golf25565 reviewed (changes requested):
Only the biggest concerns

foaw committed:
refactor(data-gen): accept broader variety of JSON sources

foaw commented on crates/data-generator/src/blocks/mod.rs:
```diff
@@ -0,0 +1,29 @@
+mod model;
+mod output;
+
+#[derive(Debug, Error)]
+pub enum GenerateError {
+    #[error("I/O operation failed")]
+    Io(#[from] IoError),
+    #[error("JSON parsing failed")]
+    Parse(#[from] JsonParseError),
+}
```
This is about as vague as it gets. Replace both messages with `#[error(transparent)]`? @rjacraft/developer

foaw committed:
feat(data-gen): support user-defined paths to `blocks.json`

foaw committed:
docs(data-gen): write descriptions to crates

foaw committed:
refactor(data-gen): rewrite generation of `enum Property`

foaw committed:
refactor(data-gen): rewrite generation of `impl TryFrom<u8> for IntProperty`

foaw committed:
refactor(data-gen): changes not worth a commit

  * indexmap.get instead of iterating through entries

  * don't default the default state ID to 0; split iteration through
    known block states into two steps: search for the default state &
    conversion into the IR

  * shorten local variables and fields, while preserving the meaning

  * add separating commas at token "unpacking", not when preparing

  * avoid using references to references that originated from oversight

foaw committed:
refactor(data-gen): rewrite generation of `BlockStateInst`

foaw committed:
refactor(data-gen): rewrite generation of `BlockStruct`
